keela-widgets/AravisController.cpp

```diff
@@ -0,0 +1,324 @@
+}
+
+std::pair<double, double> AravisController::get_gain_range() const {
+	if(aravis_camera == nullptr) {
+		spdlog::warn("Gain control not supported");
+
+		auto nan = std::numeric_limits<double>::quiet_NaN();
+		return {nan, nan};
+	}
+
+	spdlog::debug("Querying gain range from camera hardware via ArvCamera object");
+
+	double min_gain, max_gain;
+	GError *error = nullptr;
+	// Query the actual hardware gain limits
+	arv_camera_get_gain_bounds(aravis_camera, &min_gain, &max_gain, &error);
+	if(error == nullptr) {
+		spdlog::info("Queried hardware gain range from camera: {:.1f} to {:.1f} dB", min_gain, max_gain);
+	} else {
+		spdlog::warn("Error querying gain range from camera: {}", error->message);
+		g_error_free(error);
+	}
+
+	return {min_gain, max_gain};
+}
+
+double AravisController::get_gain() const {
+	GError *error = nullptr;
+	gdouble gain = arv_camera_get_gain(aravis_camera, &error);
+
+	if(error != nullptr) {
+		spdlog::error("Error getting gain from camera: {}", error->message);
+		g_error_free(error);
+		return std::numeric_limits<double>::quiet_NaN();
+	}
+
+	return gain;
+}
+
+std::pair<double, double> AravisController::get_exposure_time_range() const {
+	if(aravis_camera == nullptr) {
+		auto nan = std::numeric_limits<double>::quiet_NaN();
+		return {nan, nan};
+	}
+
+	spdlog::debug("Querying exposure time range from camera hardware via ArvCamera object");
+
+	double min_exposure, max_exposure;
+	GError *error = nullptr;
+	// Query the actual hardware exposure time limits
+	arv_camera_get_exposure_time_bounds(aravis_camera, &min_exposure, &max_exposure, &error);
+	if(error == nullptr) {
+		spdlog::info("Queried hardware exposure time range from camera: {:.1f} to {:.1f} us", min_exposure,
+		             max_exposure);
+	} else {
+		spdlog::warn("Error querying exposure time range from camera: {}", error->message);
+		g_error_free(error);
+	}
+
+	return {min_exposure, max_exposure};
+}
+
+double AravisController::get_exposure_time() const {
+	if(aravis_camera == nullptr) {
+		spdlog::warn("No ArvCamera available - cannot get exposure time");
+		return std::numeric_limits<double>::quiet_NaN();
+	}
+
+	spdlog::debug("Querying current exposure time from camera hardware via ArvCamera object");
+
+	GError *error = nullptr;
+	gint exposure_time = arv_camera_get_exposure_time(aravis_camera, &error);
+
+	if(error != nullptr) {
+		spdlog::error("Error getting exposure time from camera: {}", error->message);
+		g_error_free(error);
+		return std::numeric_limits<double>::quiet_NaN();
+	}
+
+	spdlog::info("Queried current camera exposure time: {} us", exposure_time);
+	return exposure_time;
+}
+
+bool AravisController::supports_hardware_binning() const {
+	return arv_camera_is_binning_available(aravis_camera, nullptr);
+}
+
+std::tuple<int, int, int, int> AravisController::get_binning_bounds() const {
+	if(aravis_camera == nullptr) {
+		spdlog::warn("Binning control not supported");
+		auto nan = std::numeric_limits<int>::quiet_NaN();
```

Comment:
NaN is only defined for floating point types

Apparently the implementation of this method for integral types simply returns 0
https://en.cppreference.com/w/cpp/types/numeric_limits/quiet_NaN